processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,189 @@
+
+
+  /**
+   * Given an encoded value, return its full representation.
+   *
+   * Using the example in the class description:
+   *   getActualValue(0, true) would return "Apple"
+   *   getActualValue(0, false) would return "Hello"
+   *
+   * @param intermediateValue The encoded value to be converted
+   * @param idSorted true, if the encoded value was sorted by actual value
+   *                 false, if the encoded value was ordered by time of ingestion
+   *
+   * @return The actual value associated with the encoded intermediateValue
+   *
+   */
+  public ActualType getActualValue(EncodedType intermediateValue, boolean idSorted);
+
+
+  /**
+   * Given an actual value, return its encoded form.
+   *
+   * Using the example in the class description:
+   *   getEncodedValue("Apple", true) would return 0
+   *   getEncodedValue("Apple", false) would return 2
+   *
+   * @param fullValue Actual value to be converted.
+   * @param idSorted  true, if return value should be an encoded value ordered by associated actual value
+   *                  false, if return value should be an encoded value ordered by time of ingestion
+   * @return An encoded value associated with fullValue
+   */
+  public EncodedType getEncodedValue(ActualType fullValue, boolean idSorted);
+
+
+  /**
+   * Given an encoded value that was ordered by time of ingestion, return the equivalent
+   * encoded value ordered by associated actual value.
+   *
+   * Using the example in the class description:
+   *   getSortedEncodedValueFromUnsorted(0) would return 2
+   *
+   * @param unsortedIntermediateValue value to convert
+   * @return converted value
+   */
+  public EncodedType getSortedEncodedValueFromUnsorted(EncodedType unsortedIntermediateValue);
+
+
+  /**
+   * Given an encoded value that was ordered by associated actual value, return the equivalent
+   * encoded value ordered by time of ingestion.
+   *
+   * Using the example in the class description:
+   *   getUnsortedEncodedValueFromSorted(2) would return 0
+   *
+   * @param sortedIntermediateValue value to convert
+   * @return converted value
+   */
+  public EncodedType getUnsortedEncodedValueFromSorted(EncodedType sortedIntermediateValue);
+
+
+  /**
+   * Returns an indexed structure of this dimension's sorted actual values.
+   * The integer IDs represent the ordering of the sorted values.
+   *
+   * Using the example in the class description:
+   *  "Apple"=0,
+   *  "Hello"=1,
+   *  "World"=2
+   *
+   * @return Sorted index of actual values
+   */
+  public Indexed<ActualType> getSortedIndexedValues();
+
+
+  /**
+   * Get the minimum dimension value seen by this indexer.
+   *
+   * NOTE:
+   * On an in-memory segment (IncrementaIndex), we can determine min/max values by looking at the stream of
+   * row values seen in calls to processRowValsToIndexKey().
+   *
+   * However, on a disk-backed segment (QueryableIndex), the numeric dimensions do not currently have any
```

Comment:
The description said that numeric dimensions don't exist yet in the PR, so of course they don't have the supporting structures :P.

That said, given that these interfaces are being introduced in order to allow the numeric dimensions (when they are created) to control how they are serialized, wouldn't it make sense for those implementations to also store a min/max if that is important?

It seems like it would be better than creating weird semantics that make it seem like there is only a single value in the dimension: `null`